extensions/nanoarrow_ipc/src/apps/integration_test_util.cc

```diff
@@ -0,0 +1,219 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cerrno>
+#include <deque>
+#include <iostream>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+#include "nanoarrow/nanoarrow_ipc.hpp"
+#include "nanoarrow/nanoarrow_testing.hpp"
+
+void print_help() {
+  std::cerr << "nanoarrow version " << ArrowNanoarrowVersion() << "\n";
+  std::cerr << "  Usage: integration_test_util convert\n";
+  std::cerr << "           --from [json|ipc] [file or -]\n";
+  std::cerr << "           [[--to [json] [-]] OR [--check [json|ipc] [file or -]]]\n";
+}
+
+class ArgumentParser {
```

Comment:
Not saying you should change this, but there are multiple header-only C++ libraries available for argument parsing such as https://github.com/jarro2783/cxxopts , https://github.com/jermp/cmd_line_parser , https://github.com/tanakh/cmdline or https://github.com/p-ranav/argparse .

You could also write this in Python and plug into the underlying C code using `ctypes` or `cffi`. This would allow yanking the separate shell script.